File: csharp/src/DarkStrata.CredentialCheck/Exceptions/ApiException.cs
using System;

namespace DarkStrata.CredentialCheck;

/// <summary>
/// Exception thrown when an API request fails.
/// </summary>
public sealed class ApiException : DarkStrataException
{
    /// <summary>
    /// Response body from the API (if available).
    /// </summary>
    public string? ResponseBody { get; }

    /// <summary>
    /// Creates a new ApiException.
    /// </summary>
    /// <param name="message">The error message.</param>
    /// <param name="statusCode">The HTTP status code.</param>
    /// <param name="responseBody">The response body (if available).</param>
    /// <param name="isRetryable">Whether this error is retryable.</param>
    /// <param name="innerException">The inner exception (if any).</param>
    public ApiException(
        string message,
        int statusCode,
        string? responseBody = null,
        bool isRetryable = false,
        Exception? innerException = null)
        : base(message, ErrorCode.ApiError, statusCode, isRetryable, innerException)
    {
        ResponseBody = responseBody;
    }
}

File: sdks/java/src/main/java/io/darkstrata/credentialcheck/exception/ApiException.java
package io.darkstrata.credentialcheck.exception;

/**
 * Exception thrown when the API returns an error response.
 */
public class ApiException extends DarkStrataException {

    private static final int[] RETRYABLE_STATUS_CODES = {408, 429, 500, 502, 503, 504};

    private final String responseBody;

    public ApiException(String message, int statusCode) {
        this(message, statusCode, null, null);
    }

    public ApiException(String message, int statusCode, String responseBody) {
        this(message, statusCode, responseBody, null);
    }

    public ApiException(String message, int statusCode, String responseBody, Throwable cause) {
        super(message, ErrorCode.API_ERROR, statusCode, isRetryableStatusCode(statusCode), cause);
        this.responseBody = responseBody;
    }

    /**
     * Get the response body from the API, if available.
     */
    public String getResponseBody() {
        return responseBody;
    }

    private static boolean isRetryableStatusCode(int statusCode) {
        for (int retryable : RETRYABLE_STATUS_CODES) {
            if (statusCode == retryable) {
                return true;
            }
        }
        return false;
    }
}
